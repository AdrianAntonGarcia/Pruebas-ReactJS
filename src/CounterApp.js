import React, { Fragment, useState } from "react";
import PropTypes from "prop-types";

const CounterApp = ({ value = 10 }) => {
	const [counter, setCounter] = useState(value);

	// handleAdd
	const handleAdd = (e) => {
		// No debemos mutar el state directamente
		// setCounter(counter + 1);
		/**
		 * Si no tuvieramos acceso a la variable counter, se puede recuperar así
		 * con el parámetro de entrada
		 */
		setCounter((c) => c + 1);
	};

	// handleSubstract
	const handleSubstract = () => {
		setCounter(counter - 1);
	};

	// handleReset
	const handleReset = () => {
		setCounter(value);
	};

	return (
		<Fragment>
			<h1>CounterApp</h1>
			<h2> {counter} </h2>
			{/* <button onClick={(e) => {handleAdd(e)}}>+1</button> 
			Es lo mismo que poner esto, el argumento que viene del onclick
			se pasa automáticamente como primer parámetro de la función. Igual
			que lo que haciamos con el console.log a secas.
			Al quitar los parentesis se ejecuta cuando damos el click, si pusieramos
			los parentesis se ejecutaría al cargar.
			Así pasamos como referencia la función para cuándo se haga click se ejecute
			*/}
			<button onClick={handleAdd}>+1</button>
			<button onClick={handleReset}>Reset</button>
			<button onClick={handleSubstract}>-1</button>
		</Fragment>
	);
};

CounterApp.propTypes = {
	value: PropTypes.number,
};

// CounterApp.defaultProps = {
// 	value: 10,
// };
export default CounterApp;
