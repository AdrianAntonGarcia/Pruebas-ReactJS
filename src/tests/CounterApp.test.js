import "@testing-library/dom";
import CounterApp from "../CounterApp";
import { shallow } from "enzyme";
import React from "react";

describe("Pruebas del componente CounterApp", () => {
	let wrapper = shallow(<CounterApp />);

	/**
	 * Función que se ejecuta antes de cada test,
	 * Reinicializamos el componente antes de cada test.
	 */
	beforeEach(() => {
		wrapper = shallow(<CounterApp />);
	});

	test("Debe de mostrar <CounterApp/> correctamente en sus valores por defecto", () => {
		expect(wrapper).toMatchSnapshot();
	});

	test("Debe de mostrar el valor 100 cuando se pasa ese prop", () => {
		const value = 100;
		const wrapper = shallow(<CounterApp value={value} />);

		const valueText = wrapper.find("h2").text().trim();
		expect(valueText).toBe("100");
	});

	test("Debe de incrementar con el botón +1", () => {
		/**
		 * Con el simulate, podemos lanzar un evento sobre el botón,
		 * el segundo parámetro es el objeto con los valores que podemos
		 * pasar al evento
		 */
		wrapper.find("button").at(0).simulate("click", {});

		const counterText = wrapper.find("h2").text().trim();

		expect(counterText).toBe("11");
	});

	test("Debe de decrementar con el botón -1", () => {
		/**
		 * Con el simulate, podemos lanzar un evento sobre el botón,
		 * el segundo parámetro es el objeto con los valores que podemos
		 * pasar al evento
		 */
		wrapper.find("button").at(2).simulate("click", {});
		const counterText = wrapper.find("h2").text().trim();
		expect(counterText).toBe("9");
	});

	test("Debe de colocar el valor por defecto con el botón de reset", () => {
		const wrapper = shallow(<CounterApp value={105} />);
		wrapper.find("button").at(0).simulate("click", {});
		wrapper.find("button").at(1).simulate("click", {});
		const counterText = wrapper.find("h2").text().trim();
		expect(counterText).toBe("105");
	});
});
