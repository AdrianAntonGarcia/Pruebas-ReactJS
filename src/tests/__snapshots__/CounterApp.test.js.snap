// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente CounterApp Debe de mostrar <CounterApp/> correctamente en sus valores por defecto 1`] = `
<Fragment>
  <h1>
    CounterApp
  </h1>
  <h2>
     
    10
     
  </h2>
  <button
    onClick={[Function]}
  >
    +1
  </button>
  <button
    onClick={[Function]}
  >
    Reset
  </button>
  <button
    onClick={[Function]}
  >
    -1
  </button>
</Fragment>
`;
